Writeup SSO pages (#38)
directus/docs: content/4.auth/7.sso/2.seamless.md

endpoint either via REST API or via SDK in order to have a session cookie or an `access_token`. Here are some
examples:

::tabs
	::tab{label="Via REST API / Fetch"}

    ```js
    await fetch("https://directus.myserver.com/auth/refresh", {
    	method: "POST",
    	credentials: "include",
    	headers: {
    		Accept: "application/json",
    		"Content-Type": "application/json",
    	},
    	body: JSON.stringify({ mode: "session" }),
    });
    ```

    In the above, `credentials` is required in order to send the refresh/session token cookie. This is using `'session'` mode, but it can also be 'cookie' or 'json'.
	::

	::tab{label="Via SDK in `session` Authentication Mode"}

    ```js
    import { createDirectus, authentication } from "@directus/sdk";

    const client = createDirectus("https://directus.myserver.com").with(
    	authentication("session", { credentials: "include" })
    );

    await client.refresh();
    ```
	::
::

::callout{type="info" title="Redirect Allow List"}

To allow Directus to redirect to external domains like `https://client.myserver.com/` used above, you'll need to include
it in the `AUTH_<PROVIDER>_REDIRECT_ALLOW_LIST` security setting.

::

> ## Testing Seamless SSO locally

Reviewer: ```suggestion
## Testing Seamless SSO Locally
```

The above `REFRESH_TOKEN_*` configuration will likely fail for local testing, as usually Directus won't be served under
a valid SSL certificate, which is a requirement for "Secure" cookies. Instead, for local testing purposes (and local
testing purposes only), the following configuration can be used:

**Authentication Mode: session**

```sh
SESSION_COOKIE_SECURE="false"
SESSION_COOKIE_SAME_SITE="lax"
```

**Authentication Mode: cookie**

```sh
REFRESH_TOKEN_COOKIE_SECURE="false"
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
```

> Note that no `REFRESH_TOKEN_COOKIE_DOMAIN` or `SESSION_COOKIE_DOMAIN` value is set.

Reviewer: ````suggestion
Locally, Directus won't be served under
a valid SSL certificate, which is a requirement for "Secure" cookies. Instead, for local testing purposes (and local
testing purposes only), the following configuration can be used when using session authentication:

```sh
SESSION_COOKIE_SECURE="false"
SESSION_COOKIE_SAME_SITE="lax"

```
Note that no  `SESSION_COOKIE_DOMAIN` value is set.
````

---

> # Seamless SSO

Reviewer: Description please

While sometimes you want your users to directly have access to the Directus project, in other cases you may need to
fetch private data from Directus in your client using external providers. For such cases, a special configuration is

> required to work across domains.

Reviewer: ```suggestion
While sometimes you want your users to directly have access to the Directus project, in other cases you may need to
fetch private data from Directus in your application using external providers. 
```

> For such cases, a special configuration is required to work across domains.

This is implied.

## Implementing Seamless SSO

> Setup an external provider. You'll find some examples further down below. Allow the cookie to be accessible across domains. There are two authentication mechanisms for this.

Reviewer: ```suggestion
Set up an external provider. You'll find some examples further down below. Allow the cookie to be accessible across domains. There are two authentication mechanisms for this.
```

I'm not too sure about this opener. I don't think it's clear.

**Authentication Mode: session**

```sh
AUTH_<PROVIDER>_MODE="session"
SESSION_COOKIE_DOMAIN="XXXX"
SESSION_COOKIE_SECURE="true"
SESSION_COOKIE_SAME_SITE="None"
```

> **Authentication Mode: cookie (legacy)**

Reviewer: If it's legacy, I'd vote we remove it?

Author: I'm so glad you agree, gonna joyfully remove

```sh
AUTH_<PROVIDER>_MODE="cookie"
REFRESH_TOKEN_COOKIE_DOMAIN="XXXX" #
REFRESH_TOKEN_COOKIE_SECURE="true"
REFRESH_TOKEN_COOKIE_SAME_SITE="None"
```

Replace XXXX for either of these modes with the domain of your Directus instance. For example "directus.myserver.com"

On your client, the login button should conform to the following format:

```html
<a
	href="https://directus.myserver.com/auth/login/google?redirect=https://client.myserver.com/login"
	>Login</a
>
```

In the above, `https://directus.myserver.com` should be the address of your Directus instance, while `https://client.myserver.com/login` should be the address of your client application. The `/login` path is not necessary, but helps to separate concerns.

> 

Reviewer: ```suggestion
Replace `https://directus.myserver.com` with the address of your Directus instance, while `https://client.myserver.com/login` should be the address of your client application. The `/login` path is not necessary, but helps to separate concerns.

```

Another example of sentence redundant openers

endpoint either via REST API or via SDK in order to have a session cookie or an `access_token`. Here are some
examples:

**Via REST API / Fetch**

```js
await fetch("https://directus.myserver.com/auth/refresh", {
	method: "POST",
	credentials: "include",
	headers: {
		Accept: "application/json",
		"Content-Type": "application/json",
	},
	body: JSON.stringify({ mode: "session" }),
});
```

In the above, `credentials` is required in order to send the refresh/session token cookie. This is using `'session'` mode, but it can also be 'cookie' or 'json'.

**Via SDK in `session` Authentication Mode**

```js
import { createDirectus, authentication } from "@directus/sdk";

const client = createDirectus("https://directus.myserver.com").with(
	authentication("session", { credentials: "include" })
);

await client.refresh();
```

**Via SDK in Legacy `cookie` Authentication Mode**

```js
import { createDirectus, authentication } from "@directus/sdk";

const client = createDirectus("https://directus.myserver.com").with(
	authentication("cookie", { credentials: "include" })
);

await client.refresh();
```

> 

Reviewer: Whichever snippets are kept - whether we remove cookie or not - put this in a code tab. Look at the authentication partial for an example.

::callout{type="info" title="Redirect Allow List"}

To allow Directus to redirect to external domains like `https://client.myserver.com/` used above, you'll need to include
it in the `AUTH_<PROVIDER>_REDIRECT_ALLOW_LIST` security setting.

::

## Testing Seamless SSO locally

The above `REFRESH_TOKEN_*` configuration will likely fail for local testing, as usually Directus won't be served under
a valid SSL certificate, which is a requirement for "Secure" cookies. Instead, for local testing purposes (and local
testing purposes only), the following configuration can be used:

**Authentication Mode: session**

```sh
SESSION_COOKIE_SECURE="false"
SESSION_COOKIE_SAME_SITE="lax"
```

**Authentication Mode: cookie**

```sh
REFRESH_TOKEN_COOKIE_SECURE="false"
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
```

Note that no `REFRESH_TOKEN_COOKIE_DOMAIN` or `SESSION_COOKIE_DOMAIN` value is set.

> ::callout{type="warning" title="Disabling secured cookies"}

Reviewer: ```suggestion
::callout{type="warning" title="Disabling Secured Cookies"}
```
